numbat/src/typechecker.rs

```diff
@@ -487,134 +517,182 @@ impl TypeChecker {
-                                    | typed_ast::BinaryOperator::GreaterThan
-                                    | typed_ast::BinaryOperator::LessOrEqual
-                                    | typed_ast::BinaryOperator::GreaterOrEqual
-                                    | typed_ast::BinaryOperator::Equal
-                                    | typed_ast::BinaryOperator::NotEqual => "comparison".into(),
-                                    typed_ast::BinaryOperator::LogicalAnd => "and".into(),
-                                    typed_ast::BinaryOperator::LogicalOr => "or".into(),
-                                },
-                                span_expected: lhs.full_span(),
-                                expected_name: " left hand side",
-                                expected_dimensions: self
-                                    .registry
-                                    .get_derived_entry_names_for(&lhs_type),
-                                expected_type: lhs_type,
-                                span_actual: rhs.full_span(),
-                                actual_name: "right hand side",
-                                actual_name_for_fix: "expression on the right hand side",
-                                actual_dimensions: self
-                                    .registry
-                                    .get_derived_entry_names_for(&rhs_type),
-                                actual_type: rhs_type,
-                            },
-                        ))
+                // DateTime types need special handling here, since they're not scalars with dimensions, yet some select binary operators can be applied to them
+                // TODO how to better handle all the operations we want to support with date
+                if lhs_checked.get_type() == Type::DateTime {
+                    let rhs_is_time = dtype(&rhs_checked)
+                        .ok()
+                        .map(|t| t.is_time_dimension(&self.registry))
+                        .unwrap_or(false);
+                    let rhs_is_datetime = rhs_checked.get_type() == Type::DateTime;
+
+                    if *op == BinaryOperator::Sub && rhs_is_datetime {
+                        // TODO error handling
+                        let time = self
+                            .registry
+                            .get_base_representation_for_name("Time")
+                            .unwrap();
+
+                        typed_ast::Expression::BinaryOperatorForDate(
+                            *span_op,
+                            *op,
+                            Box::new(lhs_checked),
+                            Box::new(rhs_checked),
+                            Type::Dimension(time),
+                            true,
+                        )
+                    } else if *op == BinaryOperator::Add
+                        || *op == BinaryOperator::Sub && rhs_is_time
```

Comment:
This should probably be
```suggestion
                    } else if (*op == BinaryOperator::Add
                        || *op == BinaryOperator::Sub) && rhs_is_time
```
because addition is currently not properly type-checked (try `now() + 1 meter`).